walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,156 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return slots;
+    }
+
+    slots.add(TimeRange.WHOLE_DAY);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, slots, true);
+      if (optionalSlots.isEmpty() && !request.getAttendees().isEmpty()) {
+        return findTimeSlots(eventsArr, request, slots, false);
+      }
+      return optionalSlots;
+    } else {
+      return findTimeSlots(eventsArr, request, slots, false);
+    }
+  }
+
+  /*
+   * Sort events by the start of their TimeRanges.
+   */
+  private ArrayList<Event> sortEventsByStart(Collection<Event> events) {
+    ArrayList<Event> eventsArr = new ArrayList<>(events);
+    Collections.sort(eventsArr, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Returns a list with all the available time slots when the requested meeting
+   * can take place.
+   */
+  private ArrayList<TimeRange> findTimeSlots(Collection<Event> events, MeetingRequest request,
+    ArrayList<TimeRange> slots, boolean hasOptionalAttendees) {
```

Comment:
i'm having trouble understanding what slots and hasOptionalAttendees mean, it looks like slots just has TimeRange.WHOLE_DAY when this is first called?  Does hasOptionalAttendees mean findTimeSlots should take into account optional attendees and only return time slots that work with those optional attendees?

It doesn't look like this is ever called without slots equaling WHOLE_DAY so can this just be a local variable?

Reply:
When I write a method's Javadoc comment, I try to make sure that each parameter is mentioned in the comment in a way that describes its meaning or use, or if I think its meaning is super obvious, I double-check that belief before leaving it out. I think that a Javadoc that refers to all of these parameters could help explain how the pieces fit together.

Reply:
ah I didn't realize that slots is always the same when findTimeSlots() is called -- I'm moving the declaration of slots within findTimeSlots and renaming hasOptionalAttendees to includeOptionalAttendees.